ddtrace/tracer/abandoned_spans.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2016 Datadog, Inc.
+
+package tracer
+
+import (
+	"container/list"
+	"fmt"
+	"strings"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+var tickerInterval = time.Minute
+var logSize = 9000
+
+// reportAbandonedSpans periodically finds and reports potentially abandoned
+// spans that are older than the given interval. These spans are stored in a
+// bucketed linked list, sorted by their `Start` time.
+func (t *tracer) reportAbandonedSpans(interval time.Duration) {
+	tick := time.NewTicker(tickerInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			logAbandonedSpans(t.abandonedSpans, true, interval)
+		case s := <-t.cIn:
+			bNode := t.abandonedSpans.Front()
+			if bNode == nil {
+				b := list.New()
+				b.PushBack(s)
+				t.abandonedSpans.PushBack(b)
+				break
+			}
+			for bNode != nil {
+				bucket, bOk := bNode.Value.(*list.List)
+				if !bOk {
+					bNode = bNode.Next()
+					continue
+				}
+				sNode := bucket.Front()
+				if sNode == nil {
+					bNode = bNode.Next()
+					continue
+				}
+				sp, sOk := sNode.Value.(*span)
+				if sOk && sp != nil && s.Start-sp.Start <= interval.Nanoseconds() {
+					bucket.PushBack(s)
+					break
+				}
+				bNode = bNode.Next()
+			}
+			if bNode != nil {
+				break
+			}
+			b := list.New()
+			b.PushBack(s)
+			t.abandonedSpans.PushBack(b)
+		case s := <-t.cOut:
+			for node := t.abandonedSpans.Front(); node != nil; node = node.Next() {
+				bucket, ok := node.Value.(*list.List)
+				if !ok || bucket.Front() == nil {
+					continue
+				}
+				spNode := bucket.Front()
+				sp, ok := spNode.Value.(*span)
+				if !ok {
+					continue
+				}
+				if s.Start-sp.Start <= interval.Nanoseconds() {
+					bucket.Remove(spNode)
+					if bucket.Front() == nil {
+						t.abandonedSpans.Remove(node)
+					}
+					break
+				}
+			}
+		case <-t.stop:
+			logAbandonedSpans(t.abandonedSpans, false, interval)
+			return
+		}
+	}
+}
+
+// logAbandonedSpans returns a string containing potentially abandoned spans. If `filter` is true,
+// it will only return spans that are older than the provided time `interval`. If false,
+// it will return all unfinished spans.
+func logAbandonedSpans(l *list.List, filter bool, interval time.Duration) {
+	var sb strings.Builder
+	nowTime := now()
+	spanCount := 0
+	truncated := false
+
+	for bucketNode := l.Front(); bucketNode != nil; bucketNode = bucketNode.Next() {
+		bucket, ok := bucketNode.Value.(*list.List)
+		if !ok || bucket == nil {
+			continue
+		}
+
+		// since spans are bucketed by time, finding a bucket that is newer
+		// than the allowed time interval means that all spans in this bucket
+		// and future buckets will be younger than `interval`, and thus aren't
+		// worth checking.
+		if filter {
+			spanNode := bucket.Front()
+			if spanNode == nil {
+				continue
+			}
+			sp, ok := spanNode.Value.(*span)
+			if !ok || sp == nil {
+				continue
+			}
+			if nowTime-sp.Start < interval.Nanoseconds() {
+				continue
+			}
```

Comment:
Couldn't this skip a bucket that has some spans that are too old, and some that aren't? I'm imagining this scenario:

interval=10
nowTime=24

bucket [10, 20) = {19, 10, 11, 18}
bucket [20, 30) = {21, 22, 20}

If we get here and look at the first bucket, then 24-19=5 < 10, so that entire bucket will be skipped. But in reality, we want to make sure both 10 and 11 are printed.

Or am I misunderstanding the logic here? To be fair, I'm assuming a list sorting that maybe isn't right. (either way, let's make sure there's a test for this case)

Reply:
That's a fair point. I don't think this sorting should ever happen (since spans are added to the list as they are created, so logically they should be sorted by their start time), and I'm not sure what the expected behavior here _should_ be. I guess we would expect the logs to be incorrect, then, since it's not checking these buckets for potential spans... 👀